client/network/src/request_responses.rs

```diff
@@ -345,148 +351,209 @@ impl NetworkBehaviour for RequestResponsesBehaviour {
-						} => {
-							let out = Event::RequestFinished {
-								request_id,
-								result: response.map_err(|()| RequestFailure::Refused),
-							};
-							return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
-						}
+					// Received a response from a remote to one of our requests.
+					RequestResponseEvent::Message {
+						message:
+						RequestResponseMessage::Response {
+							request_id,
+							response,
+						},
+						..
+					} => {
+						let out = Event::RequestFinished {
+							request_id,
+							result: response.map_err(|()| RequestFailure::Refused),
+						};
+						return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+					}
 
-						// One of our requests has failed.
-						RequestResponseEvent::OutboundFailure {
+					// One of our requests has failed.
+					RequestResponseEvent::OutboundFailure {
+						request_id,
+						error,
+						..
+					} => {
+						let out = Event::RequestFinished {
 							request_id,
-							error,
-							..
-						} => {
-							let out = Event::RequestFinished {
-								request_id,
-								result: Err(RequestFailure::Network(error)),
-							};
-							return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
-						}
+							result: Err(RequestFailure::Network(error)),
+						};
+						return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+					}
 
-						// Remote has tried to send a request but failed.
-						RequestResponseEvent::InboundFailure { peer, error, .. } => {
-							let out = Event::InboundRequest {
-								peer,
-								protocol: protocol.clone(),
-								result: Err(ResponseFailure::Network(error)),
-							};
-							return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+					// Remote has tried to send a request but failed.
+					RequestResponseEvent::InboundFailure { request_id, peer, error, .. } => {
+						if self.pending_responses_arrival_time.remove(&request_id).is_none() {
+							debug_assert!(
+								false,
+								"Expected to find start time for failed response.",
+							)
+						}
+						let out = Event::InboundRequest {
+							peer,
+							protocol: protocol.clone(),
+							result: Err(ResponseFailure::Network(error)),
+						};
+						return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+					}
+					RequestResponseEvent::ResponseSent { request_id, peer } => {
+						match self.pending_responses_arrival_time.remove(&request_id) {
+							Some(arrival) => {
+								let out = Event::InboundRequest {
+									peer,
+									protocol: protocol.clone(),
+									result: Ok(arrival.elapsed()),
+								};
+								return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+							},
+							None => debug_assert!(
+								false,
+								"Expected to find start time for sent response.",
+							),
 						}
-					};
-				}
-			}
 
-			break Poll::Pending;
+					}
+				};
+			}
 		}
+
+		// Sending a response can fail silently if the connection to the peer closed after calling
```

Comment:
Its not quite silent, as this can only happen if the connection suddenly closes, but since we do not yet expose connection IDs from `libp2p-request-response` (I think that should eventually happen), you probably cannot associate your arrival times with connections and clean up on connection closed. Since these arrival times are purely for statistical purposes, wouldn't it be easier and safer to e.g. use a reasonably-sized `LruCache`? Of course, you then need to account for the possibility that an entry may be gone when you look it up, resorting to a default, but it seems harmless and should happen rarely with e.g. a cache size of a few thousand entries. It would spare you the cleanup here and guarantee an upper bound on the memory usage. Doing an `O(n)` `retain` on every `poll` for this cleanup does not seem so desirable either, if there is a good amount of inbound request/response traffic. Just a thought.

Reply:
Using an LRU cache seems a lot easier to manage here plus removing the possible performance hit. https://github.com/paritytech/substrate/pull/7696/commits/2b8b0ff61983ca0143d78a4b2dfe1eb31ab24ec5 adjusts the pull request accordingly.

Its a bit hard to predict how many requests a node might handle concurrently. I set the LRU limit to `1_000` for now. We might want to increase this in the future. Thoughts?

Reply:
To back up the number `1_000` with data: None of our Kusama nor Polkadot nodes served more than 20 requests per second in the last 30 days. With that in mind I think `1_000` should serve us well without using up too much memory (`1_000` times a `u64` for the request id plus 128 bit for the `Instant` plus overhead of the `LruCache`).

Reply:
Unless this cache is used for anything else than measuring inbound request processing times, the only consideration for choosing the size seems to be up to how many concurrent inbound requests we are not willing to lose data points for the statistics. Am I missing something? I would therefore still say that it is not a good idea to [panic in debug mode](https://github.com/paritytech/substrate/blob/2b8b0ff61983ca0143d78a4b2dfe1eb31ab24ec5/client/network/src/request_responses.rs#L522-L525) when the cache dropped an entry for a request that was still active. I think a debug log message would suffice.

Reply:
> Unless this cache is used for anything else than measuring inbound request processing times, the only consideration for choosing the size seems to be up to how many concurrent inbound requests we are not willing to lose data points for the statistics.

Correct, it is only used for measuring request processing times. I initially chose `debug_assert` given that I expect the LRU cache to be large enough and given that I am worried, in case it is too small, that we miss a standard `debug` log line.

I am fine either way. https://github.com/paritytech/substrate/pull/7696/commits/7fb46fc62fccc71b7b6001307bd58ee1bf08da32 replaces the `debug_assert` with a `debug` log line.